File: 351004/Brazhalovich/Lab_1/publisher/src/main/java/org/example/newsapi/service/MarkerService.java
package org.example.newsapi.service;

import lombok.RequiredArgsConstructor;
import org.example.newsapi.dto.request.MarkerRequestTo;
import org.example.newsapi.dto.response.MarkerResponseTo;
import org.example.newsapi.entity.Marker;
import org.example.newsapi.exception.AlreadyExistsException;
import org.example.newsapi.exception.NotFoundException;
import org.example.newsapi.mapper.MarkerMapper;
import org.example.newsapi.repository.MarkerRepository;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class MarkerService {

    private final MarkerRepository markerRepository;
    private final MarkerMapper markerMapper;

    @Transactional
    @CacheEvict(value = "markers", allEntries = true)
    public MarkerResponseTo create(MarkerRequestTo request) {
        if (markerRepository.existsByName(request.getName())) {
            throw new AlreadyExistsException("Marker already exists");
        }
        Marker marker = markerMapper.toEntity(request);
        Marker saved = markerRepository.save(marker);
        return markerMapper.toDto(saved);
    }

    public Page<MarkerResponseTo> findAll(Pageable pageable) {
        return markerRepository.findAll(pageable)
                .map(markerMapper::toDto);
    }

    @Cacheable(value = "markers", key = "#id")
    public MarkerResponseTo findById(Long id) {
        return markerRepository.findById(id)
                .map(markerMapper::toDto)
                .orElseThrow(() -> new NotFoundException("Marker not found with id: " + id));
    }

    @Transactional
    @CacheEvict(value = "markers", allEntries = true)
    public MarkerResponseTo update(Long id, MarkerRequestTo request) {
        Marker marker = markerRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Marker not found with id: " + id));
        markerMapper.updateEntityFromDto(request, marker);
        return markerMapper.toDto(markerRepository.save(marker));
    }

    @Transactional
    @CacheEvict(value = "markers", allEntries = true)
    public void delete(Long id) {
        if (!markerRepository.existsById(id)) {
            throw new NotFoundException("Marker not found with id: " + id);
        }
        markerRepository.deleteById(id);
    }
}

File: 351003/brezgunov/distcomp/src/main/kotlin/com/distcomp/service/MarkerService.kt
package com.distcomp.service

import com.distcomp.dto.marker.MarkerRequestTo
import com.distcomp.dto.marker.MarkerResponseTo
import com.distcomp.exception.MarkerNotFoundException
import com.distcomp.mapper.MarkerMapper
import com.distcomp.repository.MarkerRepository
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.CachePut
import org.springframework.cache.annotation.Cacheable
import org.springframework.cache.annotation.Caching
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class MarkerService(
    val markerMapper: MarkerMapper,
    val markerRepository: MarkerRepository
) {
    @Transactional
    @CachePut(value = ["markers"], key = "#result.id")
    @CacheEvict(value = ["markers"], key = "'all'")
    fun createMarker(markerRequestTo: MarkerRequestTo): MarkerResponseTo {
        val marker = markerMapper.toMarkerEntity(markerRequestTo)
        markerRepository.save(marker)
        return markerMapper.toMarkerResponse(marker)
    }

    @Cacheable(value = ["markers"], key = "#id")
    fun readMarkerById(id: Long): MarkerResponseTo {
        val marker = markerRepository.findByIdOrNull(id)
            ?: throw MarkerNotFoundException("Marker with id $id not found")
        return markerMapper.toMarkerResponse(marker)
    }

    @Cacheable(value = ["markers"], key = "'all'")
    fun readAll(): List<MarkerResponseTo> {
        return markerRepository.findAll().map { markerMapper.toMarkerResponse(it) }
    }

    @Transactional
    @CachePut(value = ["markers"], key = "#markerId")
    @CacheEvict(value = ["markers"], key = "'all'")
    fun updateMarker(markerRequestTo: MarkerRequestTo, markerId: Long?): MarkerResponseTo {
        if (markerId == null || markerRepository.findByIdOrNull(markerId) == null) {
            throw MarkerNotFoundException("Marker with id $markerId not found")
        }
        val marker = markerMapper.toMarkerEntity(markerRequestTo)
        marker.id = markerId
        markerRepository.save(marker)
        return markerMapper.toMarkerResponse(marker)
    }

    @Transactional
    @Caching(evict = [
        CacheEvict(value = ["markers"], key = "#id"),
        CacheEvict(value = ["markers"], key = "'all'")
    ])
    fun removeMarkerById(id: Long) {
        if (markerRepository.findByIdOrNull(id) == null) {
            throw MarkerNotFoundException("Marker with id $id not found")
        }
        markerRepository.deleteById(id)
    }
}